vividus-util/src/main/java/org/vividus/util/wait/Waiter.java

```diff
@@ -28,24 +29,35 @@
 {
     private final long durationInMillis;
     private final long pollingTimeoutMillis;
+    private final int pollingRetries;
 
     public Waiter(WaitMode waitMode)
     {
         durationInMillis = waitMode.getDuration().toMillis();
         pollingTimeoutMillis = waitMode.calculatePollingTimeout(TimeUnit.MILLISECONDS);
+        pollingRetries = waitMode.getRetryTimes();
     }
 
-    public <T, E extends Exception> T wait(FailableSupplier<T, E> valueProvider, Predicate<T> stopCondition) throws E
+    public Waiter(Duration pollingTimeoutMillis, int pollingRetries)
     {
-        long endTime = System.currentTimeMillis() + durationInMillis;
+        this.durationInMillis = 0;
+        this.pollingTimeoutMillis = pollingTimeoutMillis.toMillis();
+        this.pollingRetries = pollingRetries;
+    }
 
+    public <T, E extends Exception> T wait(FailableSupplier<T, E> valueProvider, Predicate<T> stopCondition) throws E
+    {
         T value;
+        long endTime = System.currentTimeMillis() + durationInMillis;
+        int counter = 0;
         do
         {
             value = valueProvider.get();
             Sleeper.sleep(pollingTimeoutMillis, TimeUnit.MILLISECONDS);
+            counter++;
         }
-        while (!stopCondition.test(value) && System.currentTimeMillis() <= endTime);
+        while (!stopCondition.test(value) && (durationInMillis > 0
+                ? System.currentTimeMillis() <= endTime : counter < pollingRetries));
```

Comment:
I think it makes sense to have 2 implementations of waiter:
- duration based
- retry times based

in this case logic won't be mixed up